Added 1st run tooltip for the updated format bar

bummytime opened the pull request:
Fixes https://github.com/wordpress-mobile/WordPress-Editor-iOS/issues/676

This PR adds a tooltip that notes how the format bar can scroll horizontally. The tip will only display on iPhones in portrait orientation and once the tip is displayed, it will not be displayed again.

**How to test:**
* Run the app from as a fresh install on an iPhone, does the tooltip show in portrait mode when editing?
* Run the app from as a fresh install on an iPhone, does the tooltip not show in landscape mode? Exit edit mode, rotate to portrait mode, re-enter edit mode...does the tooltip appear?
* Run the app from as a fresh install on an iPad, you should not see the tooltip...ever.

**Other things to consider:**
* IME Keyboards (e.g. Japanese - Kana)
* if the user enters edit mode directly from the post list, does the tooltip work?
* if the user enters edit mode from view mode on a post, does the tooltip work?
* Works on iOS7 and iOS8
* Minimize (and maximize) the QuickType bar on the keyboard, does the tooltip display properly in either case?
* Does the tooltip appear when an external keyboard is used and the format bar is at the bottom of the screen?
* If edit mode is dismissed while the tooltip is displayed, does it disappear gracefully and immediately?

![scaled screen shot 2015-07-07 at 9 12 41 pm](https://cloud.githubusercontent.com/assets/154014/8562097/9bf51b4a-24f3-11e5-969f-7856b40c83fe.png)
![scaled screen shot 2015-07-07 at 9 13 50 pm](https://cloud.githubusercontent.com/assets/154014/8562096/9bf3e07c-24f3-11e5-9774-f5f0edd1019f.png)

@diegoreymendez  if you have a moment, can you take a look at this?

bummytime commented:
@diegoreymendez you are off the hook! :-) I got @aerych to take a peek at this.

aerych commented on WordPress/Classes/Utility/WPTooltip.h:
```diff
@@ -1,21 +1,43 @@
+ *
+ *  @return     WPToolTip   New instance of WPTooltip.
+ */
++ (instancetype)displayToolTipInView:(UIView *)view fromFrame:(CGRect)frame withText:(NSString *)text direction:(WPTooltipDirection)direction;
```
Maybe lowercase the T in tip to be consistent with the class name `WPTooltip`?  Also in the `@return` documentation.

bummytime replied:
:+1:

aerych commented on WordPress/Classes/ViewRelated/Post/WPPostViewController.m:
```diff
@@ -559,29 +583,96 @@ - (void)showUnsavedChangesAlert
+        __typeof__(self) __weak weakSelf = self;
+        NSString *tooltipText = NSLocalizedString(@"Slide for more", @"Tooltip that lets a user know they can slide the formatting toolbar displayed when the user is editing a post.");
+        dispatch_async(dispatch_get_main_queue(), ^{
+            weakSelf.formatBarToolTip = [WPTooltip displayToolTipInView:self.view fromFrame:self.keyboardRect withText:tooltipText direction:WPTooltipDirectionUp];
```
`self.view` and `self.keyboardRect` should probably use `weakSelf` instead

bummytime replied:
Gah! Of course.

aerych commented on WordPress/Classes/ViewRelated/Post/WPPostViewController.m:
```diff
@@ -903,6 +994,40 @@ + (BOOL)isNewEditorEnabled
+    // We need to adjust the vertical space a little bit for the iPad and iPhone 6 in landscape
+    CGFloat verticalOffset = 0.0;
+    if (self.isLandscape && !WPDeviceIdentification.isiPhoneSixPlus) {
+        verticalOffset = -12.0;
```
Maybe make this a const.

bummytime replied:
:+1:

aerych commented on WordPress/Classes/ViewRelated/Post/WPPostViewController.m:
```diff
@@ -903,6 +994,40 @@ + (BOOL)isNewEditorEnabled
+    
+    // We need to adjust the vertical space a little bit for the iPad and iPhone 6 in landscape
+    CGFloat verticalOffset = 0.0;
+    if (self.isLandscape && !WPDeviceIdentification.isiPhoneSixPlus) {
```
Not sure why the adjustment is needed if the tip should never be shown on the iPad, or when in Landscape orientation.  Did I misunderstand the description in the PR?

bummytime replied:
No you did not. I originally had it running on the iPad and then decided not to display the tip there...I simply did not remove this code, but will.

aerych commented on WordPress/Classes/ViewRelated/Post/WPPostViewController.m:
```diff
@@ -903,6 +994,40 @@ + (BOOL)isNewEditorEnabled
+{
+    NSDictionary* info = [aNotification userInfo];
+    CGRect rawKeyboardRect = [[info objectForKey:UIKeyboardFrameEndUserInfoKey] CGRectValue];
+    CGRect kRect = [self.view.window convertRect:rawKeyboardRect toView:self.view.window.rootViewController.view];
```
`showFormatBarOnboarding` displays the tooltip in this view controller's view, so I'm not sure why we'd want to trust a rect converted from a different view controllers view, even if the dimensions are the same.

aerych replied:
why not `[self.view convertRect:rawKeyboardRect fromView:self.view.window];`?

bummytime replied:
:+1:

aerych commented:
A quick note on the tool tip UX. The first time I launched the app the simulator was in hardware keyboard mode, and the Quick Type bar was collapsed.  I honestly thought the "swipe for more" message meant I could swipe up to see the Quick Type bar.  I'm not sure our users will catch that it means swipe horizontally to see more formatting options.

aerych commented on WordPress/Classes/ViewRelated/Post/WPPostViewController.m:
```diff
@@ -903,6 +994,40 @@ + (BOOL)isNewEditorEnabled
+
+- (void)keyboardWasShown:(NSNotification*)aNotification
+{
+    NSDictionary* info = [aNotification userInfo];
```
Code style: `NSDictionary *info`

aerych commented on WordPress/Classes/ViewRelated/Post/WPPostViewController.m:
```diff
@@ -903,6 +994,40 @@ + (BOOL)isNewEditorEnabled
+        verticalOffset = -12.0;
+    }
+    kRect.origin.y -= self.toolbarView.layer.frame.size.height + verticalOffset;
+    kRect.size.height += self.toolbarView.layer.frame.size.height;
```
Maybe use the `CGRectGetHeight` macro here?
Why use the `CALayer`'s frame instead of the view's frame?

bummytime replied:
So the view's frame does not have not correct height at the time this notification handler is called. The layer was correct so I used that.

aerych replied:
Interesting! 
Maybe add a comment?  I can see the layer being accidentally removed in the future by a well meaning edit since its typically the view's frame that's used.

bummytime replied:
:+1:

aerych commented on WordPress/Classes/ViewRelated/Post/WPPostViewController.m:
```diff
@@ -903,6 +994,40 @@ + (BOOL)isNewEditorEnabled
+    }
+}
+
+- (void)keyboardWasShown:(NSNotification*)aNotification
```
in the iPhone 5s 8.4 simulator ` +- (void)keyboardWasShown:(NSNotification*)aNotification ` is being called twice for every orientation change.  I haven't been able to suss out why that's happening.

bummytime replied:
Ok, I am not seeing that in my environment. I'll investigate.

aerych commented on WordPress/Classes/ViewRelated/Post/WPPostViewController.m:
```diff
@@ -903,6 +994,40 @@ + (BOOL)isNewEditorEnabled
+    [self showFormatBarOnboarding];
+}
+
+- (BOOL)isLandscape
```
Rather than define a new method, why not use the existing UIDevice macro where this method is called? 
`UIDeviceOrientationIsLandscape(self.interfaceOrientation)`

bummytime replied:
:+1:

bummytime commented:
@aerych @thianhlu Regarding the tooltip text, I thought we could hug it out here. @thianhlu suggested changing it to `Slide for more options`. @aerych any thoughts?

aerych commented:
I've no strong feelings about the verbiage either way.  
Random out-of-left-field-thought: Could we nudge/ease the bar when showing the tool tip to reinforce the horizontal action? If nothing else its an excuse to play with UIDynamics ;)

bummytime commented:
> Could we nudge/ease the bar when showing the tool tip

Sure! Let me try a few things.

diegoreymendez commented:
@aerych, @bummytime - Thanks for taking care of this!

diegoreymendez commented on WordPress/Classes/ViewRelated/Post/WPPostViewController.m:
```diff
@@ -289,6 +299,17 @@ - (void)viewDidLoad
+    [super viewWillAppear:animated];
+    [[NSNotificationCenter defaultCenter] addObserver:self
+                                             selector:@selector(keyboardWasShown:)
+                                                 name:UIKeyboardDidShowNotification object:nil];
```
Small detail: bring the `object:nil` parameter to the next line would improve readability.

bummytime commented:
Ok @aerych and @thianhlu. I took your latest batch of suggestions and implemented them. Video here:

https://cloudup.com/c-rINfAUdZK

Things that changed:
* Moved format bar tooltip to the right
* Changed format bar tooltip text
* Now showing the format bar tooltip when the format bar buttons are enabled (e.g. when the user taps the content body)
* Added a scrolling animation to coincide with the tooltip

Note that I am purposefully resetting the on-boarding flags after the editor screen closes for demo purposes. In the wild, this would run once per user.

thianhlu commented:
Thanks Bummy! looks great


> On Jul 13, 2015, at 7:23 PM, Matt Bumgardner <notifications@github.com> wrote:
> 
> Ok @aerych and @thianhlu. I took your latest batch of suggestions and implemented them. Video here:
> 
> https://cloudup.com/c-rINfAUdZK
> 
> Things that changed:
> 
> Moved format bar tooltip to the right
> Changed format bar tooltip text
> Now showing the format bar tooltip when the format bar buttons are enabled (e.g. when the user taps the content body)
> Added a scrolling animation to coincide with the tooltip
> Note that I am purposefully resetting the on-boarding flags after the editor screen closes for demo purposes. In the wild, this would run once per user.
> 
> —
> Reply to this email directly or view it on GitHub.
>

aerych commented:
Looks good to me @bummytime!  There are merge conflicts but otherwise :shipit:

bummytime commented:
Thank you @aerych and @thianhlu !
